### PR TITLE
Split control/SearchControl into view/controller

### PR DESCRIPTION
**Pull request by bhennion**
This PR splits control/SearchControl into a view and a controller.

Some small side-refactoring:
- searched strings are now passed by references (instead of by a succession of copies and moves)

Ready for a review.

### CODE REVIEW

File: src/core/control/Control.cpp [hunk 1]
```diff
@@ -1749,8 +1749,9 @@ auto Control::getCurrentPageNo() -> size_t {
     return 0;
 }
 
-auto Control::searchTextOnPage(string text, int p, int* occures, double* top) -> bool {
-    return getWindow()->getXournal()->searchTextOnPage(std::move(text), p, occures, top);
+auto Control::searchTextOnPage(const string& text, size_t pageNumber, size_t* occurences, double* yOfUpperMostMatch)
```
Comment (Technius) [inline src/core/control/Control.cpp:1752]:
```suggestion
auto Control::searchTextOnPage(const std::string& text, size_t pageNumber, size_t* occurences, double* yOfUpperMostMatch)
```

File: src/core/control/Control.h [hunk 2]
```diff
@@ -187,7 +187,7 @@ class Control:
 
     bool isFullscreen();
 
-    bool searchTextOnPage(std::string text, int p, int* occures, double* top);
+    bool searchTextOnPage(const std::string& text, size_t pageNumber, size_t* occurences, double* yOfUpperMostMatch);
```
Comment (Technius) [inline src/core/control/Control.h:190]: typo?
```suggestion
    bool searchTextOnPage(const std::string& text, size_t pageNumber, size_t* occurrences, double* yOfUpperMostMatch);
```

Comment (Technius) [reply src/core/control/Control.h:190]: It would also be helpful to document this method.

File: src/core/gui/SearchBar.cpp [hunk 3]
```diff
@@ -99,88 +99,86 @@ void SearchBar::searchTextChangedCallback(GtkEntry* entry, SearchBar* searchBar)
...
+    size_t currentPage = control->getCurrentPageNo();
+    size_t count = control->getDocument()->getPageCount();
     if (count < 2) {
         // Nothing to do
         return;
     }
 
     MainWindow* win = control->getWindow();
-    int x = page - 1;
+    size_t searchedPage = currentPage == 0 ? count - 1 : currentPage - 1;
```
Comment (Technius) [inline src/core/gui/SearchBar.cpp:155]: nit: this line should be moved closer to the loop.

File: src/core/gui/SearchBar.cpp [hunk 4]
```diff
@@ -99,88 +99,86 @@ void SearchBar::searchTextChangedCallback(GtkEntry* entry, SearchBar* searchBar)
...
-        x = count - 1;
-    }
-
-    double top = 0;
-    int occures = 0;
+    double yOfUpperMostMatch = 0;
+    size_t occurences = 0;
 
-    while (x != page) {
+    while (searchedPage != currentPage) {
```
Comment (Technius) [inline src/core/gui/SearchBar.cpp:166]:
```suggestion
    // Search backwards through the pages, wrapping around if needed.
    while (searchedPage != currentPage) {
```

Comment (Technius) [reply src/core/gui/SearchBar.cpp:166]: Could also simplify into a for loop, but not sure if that would be more readable.

**Review by Technius — APPROVED** [review]
LGTM. Minor nitpicks below:

File: src/core/gui/SearchBar.cpp (hunk 4 above)
Comment (bhennion) [reply src/core/gui/SearchBar.cpp]: I refactored SearchPage::searchNext and searchPrevious to remove the code duplication, and I documented the functions.
I chose not the change the behaviour,: the current page is not searched.
Those functions are called when pressing the next/previous button on the search bar. This then finds the next/previous page with matches and scrolls to said page.
I think the behaviour is a little weird (it should first navigate through the matches of the current page), but changing this is out of scope.

File: src/core/control/SearchControl.cpp [hunk 5]
```diff
@@ -9,39 +9,26 @@
...
-        cairo_stroke_preserve(cr);
-        auto applied = GdkRGBA{color.red, color.green, color.blue, 0.3};
-        gdk_cairo_set_source_rgba(cr, &applied);
-        cairo_fill(cr);
-    }
-}
-
-auto SearchControl::search(std::string text, int* occures, double* top) -> bool {
-    freeSearchResults();
+SearchControl::~SearchControl() { this->results.clear(); }
```
Comment (Technius) [inline src/core/control/SearchControl.cpp:19]: nit: won't `results` will be cleared automatically when it's destroyed?
```suggestion
SearchControl::~SearchControl() = default;
```

**Review by Technius — APPROVED** [review]

**Comment by bhennion** [comment]
Thanks for the review @Technius!
I cleaned up the commit tree.

Merging in 48 hours unless an objection is raised.

**Pull request merged**